File: lib/src/constants/temperature_constants.dart
// FILE: lib/src/constants/temperature_constants.dart
// VERSION: 1.0.0
// START_MODULE_CONTRACT
//   PURPOSE: Маппинг температуры салона в TemperatureRange и HeatSequence —
//            источник расписаний авторежима и значений слайдера порога.
//   SCOPE: пороги диапазонов, расписания warm/cool/cold/freezing/extreme,
//          pure-функции getTemperatureRange и getHeatSequence.
//   DEPENDS: none
//   LINKS: M-CONSTANTS-TEMPERATURE, V-M-CONSTANTS-TEMPERATURE
//   ROLE: RUNTIME
//   MAP_MODE: EXPORTS
// END_MODULE_CONTRACT
//
// START_MODULE_MAP
//   TemperatureRange - enum off|warm|cool|cold|freezing|extreme
//   HeatSequence - длительности уровней 3/2/1 (минуты)
//   TemperatureConstants - контейнер порогов, расписаний и pure-функций
//   temperatureThresholds - нижние границы диапазонов (порядок итерации значим)
//   sliderMin / sliderMax / sliderDivisions - параметры слайдера порога
//   sliderValues / sliderLabels - значения и подписи слайдера [-5..15]
//   defaultTemperatureThreshold - 5.0, порог авторежима по умолчанию
//   temperatureSequences - HeatSequence по диапазонам (без off)
//   getTemperatureRange - °C -> TemperatureRange по temperatureThresholds
//   getHeatSequence - °C -> HeatSequence? (null при range.off)
// END_MODULE_MAP
//
// START_CHANGE_SUMMARY
//   LAST_CHANGE: [v0.2.0 - GRACE-инициализация: добавлены MODULE_CONTRACT и MODULE_MAP]
// END_CHANGE_SUMMARY

enum TemperatureRange {
  off, // > 10°C
  warm, // 5°C до 10°C
  cool, // 0°C до 5°C
  cold, // -5°C до 0°C
  freezing, // -10°C до -5°C
  extreme, // < -10°C
}

class HeatSequence {
  final int level3Duration;
  final int level2Duration;
  final int level1Duration;

  /// Температура, при которой уровень 3 → уровень 2 (досрочно, не дожидаясь таймера)
  final double level3StepDownCelsius;
  /// Температура, при которой уровень 2 → уровень 1
  final double level2StepDownCelsius;
  /// Температура, при которой уровень 1 → 0
  final double level1StepDownCelsius;

  const HeatSequence({
    required this.level3Duration,
    required this.level2Duration,
    required this.level1Duration,
    this.level3StepDownCelsius = double.negativeInfinity,
    this.level2StepDownCelsius = double.negativeInfinity,
    this.level1StepDownCelsius = double.negativeInfinity,
  });
}

class TemperatureConstants {
  static const Map<TemperatureRange, double> temperatureThresholds = {
    TemperatureRange.off: 10.0,
    TemperatureRange.warm: 5.0,
    TemperatureRange.cool: 0.0,
    TemperatureRange.cold: -5.0,
    TemperatureRange.freezing: -10.0,
    TemperatureRange.extreme: double.negativeInfinity,
  };

  static const double sliderMin = -5.0;
  static const double sliderMax = 15.0;
  static const int sliderDivisions = 4;

  static const List<double> sliderValues = [-5, 0, 5, 10, 15];
  static const List<String> sliderLabels = ['-5°C', '0°C', '5°C', '10°C', '15°C'];

  static const double defaultTemperatureThreshold = 5.0;

  static const Map<TemperatureRange, HeatSequence> temperatureSequences = {
    TemperatureRange.warm: HeatSequence(
      level3Duration: 3,
      level2Duration: 2,
      level1Duration: 5,
      level3StepDownCelsius: 6.0,
      level2StepDownCelsius: 8.0,
      level1StepDownCelsius: 11.0,
    ), // +5°C до +10°C: 10 мин
    TemperatureRange.cool: HeatSequence(
      level3Duration: 5,
      level2Duration: 3,
      level1Duration: 7,
      level3StepDownCelsius: 2.0,
      level2StepDownCelsius: 6.0,
      level1StepDownCelsius: 9.0,
    ), // 0°C до +5°C: 15 мин
    TemperatureRange.cold: HeatSequence(
      level3Duration: 8,
      level2Duration: 5,
      level1Duration: 7,
      level3StepDownCelsius: -2.0,
      level2StepDownCelsius: 3.0,
      level1StepDownCelsius: 7.0,
    ), // -5°C до 0°C: 20 мин
    TemperatureRange.freezing: HeatSequence(
      level3Duration: 12,
      level2Duration: 7,
      level1Duration: 7,
      level3StepDownCelsius: -7.0,
      level2StepDownCelsius: -2.0,
      level1StepDownCelsius: 4.0,
    ), // -10°C до -5°C: 26 мин
    TemperatureRange.extreme: HeatSequence(
      level3Duration: 15,
      level2Duration: 10,
      level1Duration: 8,
      level3StepDownCelsius: -12.0,
      level2StepDownCelsius: -7.0,
      level1StepDownCelsius: 0.0,
    ), // < -10°C: 33 мин
  };

  static TemperatureRange getTemperatureRange(double temperature) {
    for (final entry in temperatureThresholds.entries) {
      if (temperature >= entry.value) {
        return entry.key;
      }
    }
    return TemperatureRange.extreme;
  }

  static HeatSequence? getHeatSequence(double temperature) {
    final range = getTemperatureRange(temperature);
    if (range == TemperatureRange.off) return null;
    return temperatureSequences[range];
  }
}

File: app/src/main/java/com/wt/airconditioner/TemperatureConstants.java
package com.wt.airconditioner;

/**
 * Расписания подогрева по температуре салона. Порт
 * lib/src/constants/temperature_constants.dart — значения перенесены дословно,
 * они настроены по живым поездкам и менять их при переписывании нельзя.
 *
 * Диапазон задаёт две вещи: сколько уровень держится максимум (таймер) и при
 * какой температуре можно шагнуть вниз досрочно (порог). Чем холоднее, тем
 * дольше держим и тем ниже порог.
 */
final class TemperatureConstants {

    /** Выше этой температуры подогрев не нужен вовсе. */
    static final double OFF_ABOVE_CELSIUS = 10.0;

    private TemperatureConstants() {
    }

    /**
     * Длительности уровней 3/2/1 в минутах и температуры досрочного перехода.
     * Порог по умолчанию — минус бесконечность: у пресетов досрочных переходов
     * нет, каскад идёт строго по таймеру.
     */
    static final class HeatSequence {
        final int level3Minutes;
        final int level2Minutes;
        final int level1Minutes;
        final double level3StepDownCelsius;
        final double level2StepDownCelsius;
        final double level1StepDownCelsius;

        HeatSequence(int level3Minutes, int level2Minutes, int level1Minutes) {
            this(level3Minutes, level2Minutes, level1Minutes,
                    Double.NEGATIVE_INFINITY, Double.NEGATIVE_INFINITY, Double.NEGATIVE_INFINITY);
        }

        HeatSequence(int level3Minutes, int level2Minutes, int level1Minutes,
                double level3StepDownCelsius, double level2StepDownCelsius,
                double level1StepDownCelsius) {
            this.level3Minutes = level3Minutes;
            this.level2Minutes = level2Minutes;
            this.level1Minutes = level1Minutes;
            this.level3StepDownCelsius = level3StepDownCelsius;
            this.level2StepDownCelsius = level2StepDownCelsius;
            this.level1StepDownCelsius = level1StepDownCelsius;
        }

        int minutesFor(int level) {
            switch (level) {
                case 3:
                    return level3Minutes;
                case 2:
                    return level2Minutes;
                case 1:
                    return level1Minutes;
                default:
                    return 0;
            }
        }

        /** Температура, при которой уровень можно снизить, не дожидаясь таймера. */
        double stepDownCelsiusFor(int level) {
            switch (level) {
                case 3:
                    return level3StepDownCelsius;
                case 2:
                    return level2StepDownCelsius;
                case 1:
                    return level1StepDownCelsius;
                default:
                    return Double.POSITIVE_INFINITY;
            }
        }

        /**
         * Сравнение планов: каскад, дошедший до нуля, перезапускается только
         * если расписание сменилось. Без этого повторное событие с той же
         * температурой снова включало бы уровень 3 — бесконечный цикл.
         */
        String planKey() {
            return level3Minutes + "," + level2Minutes + "," + level1Minutes + ","
                    + level3StepDownCelsius + "," + level2StepDownCelsius + ","
                    + level1StepDownCelsius;
        }
    }

    private static final HeatSequence WARM = new HeatSequence(3, 2, 5, 6.0, 8.0, 11.0);
    private static final HeatSequence COOL = new HeatSequence(5, 3, 7, 2.0, 6.0, 9.0);
    private static final HeatSequence COLD = new HeatSequence(8, 5, 7, -2.0, 3.0, 7.0);
    private static final HeatSequence FREEZING = new HeatSequence(12, 7, 7, -7.0, -2.0, 4.0);
    private static final HeatSequence EXTREME = new HeatSequence(15, 10, 8, -12.0, -7.0, 0.0);

    /** Расписание для температуры салона; null — теплее порога, греть не нужно. */
    static HeatSequence heatSequenceFor(double celsius) {
        if (celsius >= OFF_ABOVE_CELSIUS) {
            return null;
        }
        if (celsius >= 5.0) {
            return WARM;
        }
        if (celsius >= 0.0) {
            return COOL;
        }
        if (celsius >= -5.0) {
            return COLD;
        }
        if (celsius >= -10.0) {
            return FREEZING;
        }
        return EXTREME;
    }
}
